iree/compiler/Dialect/HAL/Target/WebGPU/WebGPUTarget.cpp

```diff
@@ -0,0 +1,270 @@
+#include "spirv-tools/libspirv.hpp"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+WebGPUTargetOptions getWebGPUTargetOptionsFromFlags() {
+  static llvm::cl::opt<bool> clDebugSymbols(
+      "iree-webgpu-debug-symbols",
+      llvm::cl::desc(
+          "Include debug information like variable names in outputs"),
+      llvm::cl::init(false));
+
+  static llvm::cl::opt<bool> clWebGPUKeepShaderModules(
+      "iree-webgpu-keep-shader-modules",
+      llvm::cl::desc("Save shader modules to disk separately"),
+      llvm::cl::init(false));
+
+  WebGPUTargetOptions targetOptions;
+  targetOptions.keepShaderModules = clWebGPUKeepShaderModules;
+
+  return targetOptions;
+}
+
+// TODO(scotttodd): provide a proper target environment for WebGPU.
+static spirv::TargetEnvAttr getWebGPUTargetEnv(MLIRContext *context) {
+  // TODO(scotttodd): find list of SPIR-V extensions supported by WebGPU/WGSL
+  auto triple = spirv::VerCapExtAttr::get(
+      spirv::Version::V_1_0, {spirv::Capability::Shader},
+      {spirv::Extension::SPV_KHR_storage_buffer_storage_class}, context);
+  return spirv::TargetEnvAttr::get(triple, spirv::Vendor::Unknown,
+                                   spirv::DeviceType::Unknown,
+                                   spirv::TargetEnvAttr::kUnknownDeviceID,
+                                   spirv::getDefaultResourceLimits(context));
+}
+
+class WebGPUTargetBackend : public TargetBackend {
+ public:
+  WebGPUTargetBackend(WebGPUTargetOptions options)
+      : options_(std::move(options)) {}
+
+  // NOTE: we could vary this based on the options such as 'webgpu-v2'.
+  std::string name() const override { return "webgpu"; }
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Codegen::IREECodegenDialect, spirv::SPIRVDialect,
+                    gpu::GPUDialect>();
+  }
+
+  IREE::HAL::DeviceTargetAttr getDefaultDeviceTarget(
+      MLIRContext *context) const override {
+    Builder b(context);
+    SmallVector<NamedAttribute> configItems;
+
+    configItems.emplace_back(b.getIdentifier("executable_targets"),
+                             getExecutableTargets(context));
+
+    auto configAttr = b.getDictionaryAttr(configItems);
+    return IREE::HAL::DeviceTargetAttr::get(
+        context, b.getStringAttr(deviceID()), configAttr);
+  }
+
+  void buildTranslationPassPipeline(OpPassManager &passManager) override {
+    buildSPIRVCodegenPassPipeline(passManager);
+    // TODO(scotttodd): additional passes for WebGPU/WGSL
+    //                  (here or during serialization?)
+  }
+
+  LogicalResult serializeExecutable(IREE::HAL::ExecutableVariantOp variantOp,
+                                    OpBuilder &executableBuilder) override {
+    ModuleOp innerModuleOp = variantOp.getInnerModule();
+    auto spirvModuleOps = innerModuleOp.getOps<spirv::ModuleOp>();
+    if (!llvm::hasSingleElement(spirvModuleOps)) {
+      return variantOp.emitError()
```

Comment:
today - we'll want multiple once we link modules together

Reply:
part of https://github.com/google/iree/issues/7824, yeah?